google-cloud-logging/src/main/java/com/google/cloud/logging/MonitoredResourceUtil.java

```diff
@@ -144,135 +206,152 @@ public static MonitoredResource getResource(String projectId, String resourceTyp
+   */
+  private static String getK8sNamespace() {
+    String value = "";
+    try {
+      value =
+          new String(Files.readAllBytes(Paths.get(K8S_POD_NAMESPACE_PATH)), StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      // if SA token is not shared the info about namespace is unavailable
+      // allow users to define the namespace name explicitly
+      value = getter.getEnv("NAMESPACE_NAME");
+      if (value == null) {
+        value = "";
+      }
     }
-    // default Resource type
-    return Resource.Global;
-  }
-
-  private static String getAppEngineModuleId() {
-    return System.getenv("GAE_SERVICE");
+    return value;
   }
 
-  private static String getAppEngineVersionId() {
-    return System.getenv("GAE_VERSION");
+  /**
+   * Distinguish between Standard and Flexible GAE environments. There is no indicator of the
+   * environment. The path to the startup-script in the metadata attribute was selected as one of
+   * the new values that explitly mentioning "flex" and cannot be altered by user (e.g. environment
+   * variable). The method assumes that the resource type is already identified as {@link
+   * Resource.AppEngine}.
+   *
+   * @return "flex" {@link String} for the Flexible environment and "standard" for the Standard.
+   */
+  private static String getAppEngineEnvironment() {
+    String value = getter.getAttribute("instance/attributes/startup-script");
+    if (value == "/var/lib/flex/startup_script.sh") {
+      return FLEX_ENV;
+    }
+    return STD_ENV;
   }
 
-  private static String getAppEngineInstanceName() {
-    return System.getenv("GAE_INSTANCE");
+  /**
+   * Retrieves a region from the qualified region of 'projects/[PROJECT_NUMBER]/regions/[REGION]'
+   *
+   * @return region string id
+   */
+  private static String getRegion() {
+    String loc = getter.getAttribute("instance/region");
+    return loc.substring(loc.lastIndexOf('/') + 1);
   }
 
-  private static String getCloudRunLocation() {
-    String zone = MetadataConfig.getZone();
-    // for Cloud Run managed, the zone is "REGION-1"
-    // So, we need to strip the "-1" to set location to just the region
-    if (zone.endsWith("-1")) return zone.substring(0, zone.length() - 2);
-    else return zone;
+  /**
+   * Retrieves a zone from the qualified zone of 'projects/[PROJECT_NUMBER]/zones/[ZONE]'
+   *
+   * @return zone string id
+   */
+  private static String getZone() {
+    String loc = getter.getAttribute("instance/zone");
+    return loc.substring(loc.lastIndexOf('/') + 1);
   }
 
   private static List<LoggingEnhancer> createEnhancers(Resource resourceType) {
     List<LoggingEnhancer> enhancers = new ArrayList<>(2);
-    switch (resourceType) {
-        // Trace logging enhancer is supported on GAE Flex and Standard.
-      case GaeAppFlex:
+    if (resourceType == Resource.AppEngine) {
+      enhancers.add(new TraceLoggingEnhancer(APPENGINE_LABEL_PREFIX));
+      if (getAppEngineEnvironment() == FLEX_ENV) {
```

Comment:
Why the logic is changed here?
We need to handle trace in AppEngine standard as well

Reply:
Also, I didn't see a test that verifies TraceLoggingEnhancer is added

Reply:
Because the legacy logic used **two** resource types for GAE and then extra converted them into one. They also populated them with different labels which comes across the public documentation description.
End users still see **single** resource type. So the functionality is not changed but the GAE environment is identified via different condition. I am considering to remove the whole JUL handler logic from this file and review it. 
I think we will need to review the support for JUL and replace it with support of the multiple logging frameworks instead. Logback appender can be used with JUL without losing in functionality.

Reply:
Yeah, I see now that there is only gae_instance (single resource - no matter if it's gae flex or standard). Thanks for clarifying.